### DISCUSSION TITLE
k8s用 ilogtail 进行自定义文件日志收集的时候，file_log 类型如何获取到文件的高精度时间

### DISCUSSION THREAD

**Discussion by wkai886** [alibaba/ilogtail#719] [Help] [answered]
ilogtail 部署在 k8s 集群里， 通过配置文件收集 input_docker_stdout 时，获取到的 contents 里面有高景区的 _time_ ,可以将收集到的日志进行排序。
但是收集 file_log 类型的 日志时， contents 里面没有 _time_ 字段，contents 外面的 _time_ 字段不是高精度的时间。
![image](https://user-images.githubusercontent.com/41099009/226260795-10396977-62c5-4339-b231-ed31cbdd4b60.png)
![640D5505-9886-4CC3-9AA3-627FEF2429FD](https://user-images.githubusercontent.com/41099009/226260827-216fc6e0-961a-473e-b0d2-36fc0e1d5155.png)

**Comment by messixukejia** [comment, marked as answer]
可以参考https://help.aliyun.com/practice_detail/427524?spm=a2c4g.26937898.0.0.294557c6zPh8P5#p-i50-8wo-700 中的 processor_strptime使用。

如果方便欢迎社区版完善下文档，欢迎提交pr。

另外，欢迎将使用logtail的场景做下记录https://github.com/alibaba/ilogtail/issues/693

**Reply by wkai886** [reply]
这个插件我看文档是只支持 sls 输出？ 我想输出到 kafka或者 pulsar 这个插件适用不

**Reply by messixukejia** [reply]
processor_strptime是通用的，只是上面案例是sls的例子。

**Reply by wkai886** [reply]
zp-orange-testtcl-zhipin-orange-web--flemkkg.yaml: |
    enable: true
    flushers:
      - Type: flusher_stdout
        OnlyStdout: true
    processors:
      - Type: processor_strptime
        EnablePreciseTimestamp: true
        PreciseTimestampUnit: us
    inputs:
      - Type: file_log
        LogPath: /logs/
        FilePattern: log-file-test-3.log
        ContainerInfo:
          IncludeK8sLabel:
            cccc.com/app: log-test
          ExternalK8sLabelTag:
            cccc.com/app: app
            cccc.com/system: system
          ExternalEnvTag:
            cccc: group


用这个配置的时候，就不开始收集日志了，去掉中间的processors 就恢复正常。
看日志未发现明显报错

ilogtail.LOG
![image](https://user-images.githubusercontent.com/41099009/226310158-31caaf8c-c6aa-497d-a6c4-eefb1e862ac6.png)

logtail_plugin.log
![image](https://user-images.githubusercontent.com/41099009/226310333-992a8e9c-308c-4b8b-943a-dd744a3afaa5.png)

**Reply by wkai886** [reply]
更新： 加了
 processors:
  - Type: processor_strptime
    EnablePreciseTimestamp: true
    PreciseTimestampUnit: us
    Format: "%Y-%m-%d %H:%M:%S.%f"

加了 Format ，stdout 看还是没有  时间字段

**Reply by messixukejia** [reply]
![image](https://user-images.githubusercontent.com/13250519/226320241-14f99276-806a-47bb-ae7d-dc22628e3b08.png)
结合代码看下这个错误日志吧，应该是参数没配全

**Reply by wkai886** [reply]
仔细看了下文档，是我理解错了，这个时间戳是需要在实际的日志文件中提取，我以为是 和收集 docker的 stdout 一样，配置好这个之后 就会有这个字段了。 业务方的日志打印方式我们没有限制，不是所有的日志都有时间这个字段，并且时间格式也各不相同。  既然 docker的 stdout 是取的 docker包的那一层的 time来做的高精度时间， 文件这个 是不是可以 就取当前时间去作为日志的收集时间，不一定要和打印时间一直，取收集时间，我们用的 log-pilot 就是取的收集时间。

**Reply by messixukejia** [reply]
logtail原有模型只支持秒级，近期扩展的模型已经支持更高精度。但是需要适配下。
![image](https://user-images.githubusercontent.com/13250519/226350309-900300af-9eab-410f-8daa-b950c46ae71f.png)

**Reply by luckqk** [reply]
请问扩展的模型指的是什么？现在ilogtail支持更高精度了吗？

**Reply by messixukejia** [reply]
需要有些衔接的开发才行，有需要可以联系我们一起开发下。
